### ISSUE TITLE
Cannot connect to wayvnc when running on labwc with no local display connected

### ISSUE THREAD

**Issue by spl237** [any1/wayvnc#286]
- **Version:**
  v0.8-rc0-6975b25 (pios)

- Describe how to **reproduce** the problem
When running wayfire with no local HDMI display connected, it is still possible to connect to WayVNC, and some default framebuffer is used. But when running labwc instead of wayfire, while connecting to WayVNC works fine if there is a local HDMI display, if there is no local HDMI display, the connection appears briefly with a window which is all black, and a second or so later the connection is dropped with the error message "An unexpected error occurred when communicating with the server: Invalid PixelBuffer of 21845 pixels requested"

This is using the TigerVNC client.

**Comment by any1** [comment]
What happens if stop the wayvnc service and run it from terminal instead, attaching directly to the headless session? A trace log might be useful.

**Comment by any1** [comment]
Did you maybe run it in detached mode (`-D` or `--detached`)? That would give you a grey screen until it's attached using the `wayvncctl attach` command.

**Comment by spl237** [comment]
Yes, I just ran wayvnc-run.sh from /usr/sbin; that has --detached in it.

Running wayvncctl attach (which I obviously have to do via SSH, because I have no local monitor) reports "Failed to find socket path "/run/user/1000/wayvncctl": No such file or directory"

Running wayvnc-run.sh with the --detached option removed (which requires sudo to access the crypto credentials) reports that it failed to connect to WAYLAND_DISPLAY="wayland-0", but again, I am really not sure that I am doing something sensible here.

Can you please give me step-by-step instructions as to exactly what I should do to run from the terminal rather than the service and capture the relevant trace? I originally tried just adding the -Ltrace option and the pipe through tee to wayvnc-run.sh, but that didn't produce any output file.

**Comment by any1** [comment]
If you run it with `--disable-input`, does that change anything?

**Comment by Consolatis** [comment]
> * When running wayfire with no local HDMI display connected, it is still possible to connect to WayVNC, and some default framebuffer is used. But when running labwc instead of wayfire, while connecting to WayVNC works fine if there is a local HDMI display, if there is no local HDMI display, the connection appears briefly with a window which is all black, and a second or so later the connection is dropped with the error message "An unexpected error occurred when communicating with the server: Invalid PixelBuffer of 21845 pixels requested"

Labwc doesn't use any fallback headless output when there are no outputs left.
You could either start labwc in headless mode (`WLR_BACKENDS=headless`, but couldn't add local outputs anymore) or you could add a virtual output to the usual session via [VirtualOutputAdd](https://labwc.github.io/labwc-actions.5.html#entry_action_name=virtualoutputadd_output_name=value).

**Comment by any1** [comment]
Thanks, Consolatis.

We can extend `wayvnc-control.py` in the pios branch to add and remove the virtual output depending on whether there is a physical one available or not.

**Comment by spl237** [comment]
>  or you could add a virtual output to the usual session via [VirtualOutputAdd](https://labwc.github.io/labwc-actions.5.html#entry_action_name=virtualoutputadd_output_name=value).

Could labwc automatically create a virtual output if there are no real outputs detected? Is there any reason not to do that? (I am assuming this is what wayfire does anyway?)

**Comment by any1** [comment]
As far as I understand, wayfire has a dummy headless output for the purpose of always having an output available in case some client depends on there always being an output available. labwc doesn't care about that (or need it?).

I've been looking at this and the way things are implemented in wayfire, it seems that capturing of the NOOP output just works by accident. I don't think that it was meant to be used for this purpose. However, I did manage to hack it so that you can control the NOOP output via output-manager and I'll send you the patches for that soon.

**Comment by Consolatis** [comment]
> > or you could add a virtual output to the usual session via [VirtualOutputAdd](https://labwc.github.io/labwc-actions.5.html#entry_action_name=virtualoutputadd_output_name=value).
> 
> Could labwc automatically create a virtual output if there are no real outputs detected? Is there any reason not to do that?

Technically we could, yes. However we decided against that to not accidentally cover up potential crashes or other bugs that may arise when there is no output left.

> As far as I understand, wayfire has a dummy headless output for the purpose of always having an output available in case some client depends on there always being an output available. labwc doesn't care about that (or need it?).

Need it.

**Comment by any1** [comment]
@Consolatis, do you know if it's possible to execute the `VirtualOutputAdd` action on startup or if you can send it via unix socket?

**Comment by Consolatis** [comment]
> if it's possible to execute the `VirtualOutputAdd` action on startup

Currently that is not possible in a clean way. We thought about having a generic autostart config section for actions (rather than random applications) but so far we only see a use-case for the `VirtualOutputAdd` command which points to a generic autostart section not being the right move forwards. We are currently not sure how to proceed with a solution. Another option could be to add a config option that creates a virtual output by default in a disabled state so applications like wayvnc could enable it via the output management protocol or wlr-randr before connecting and disable it again when disconnecting. CC @johanmalm

In the meantime one can hack around the issue by injecting a keyboard shortcut via the virtual keyboard protocol (e.g. via `wtype` or `wlrctl` in the generic `~/.config/labwc/autostart`) but that obviously requires user configuration and is in general pretty hacky.

> or if you can send it via unix socket?

Labwc purposefully does not support any kind of IPC other than via the wayland / wlroots protocols (and `SIGHUP` to reload the config if you want to call that IPC).

**Comment by any1** [comment]
> In the meantime one can hack around the issue by injecting a keyboard shortcut via the virtual keyboard protocol (e.g. via wtype or wlrctl in the generic ~/.config/labwc/autostart or before connecting via wayvnc) but that obviously requires user configuration of the keybind and is in general pretty hacky.

That is very hacky indeed and probably not even much less of an effort than just implementing the "enable fallback virtual output" config option.

**Comment by spl237** [comment]
Agreed - automatically creating a virtual output when no real outputs are present would be by far the tidiest way of doing this.

If that behaviour is not always desirable for the purpose of debugging etc - which is fine - then let's have it as a command-line option or similar with it disabled by default. But hacks involving injecting keyboard strokes are really not the way we should be doing this!

**Comment by Consolatis** [comment]
Labwc just merged https://github.com/labwc/labwc/pull/1586 which allows enabling a fallback output by setting an env var to something like `LABWC_FALLBACK_OUTPUT=NOOP-fallback`. That should then also allow wayvnc to detect it as a headless output.

**Comment by spl237** [comment]
I've been fiddling witht this some more, and there seems to be an occasional bug whereby connecting an HDMI output after booting with no output crashes the outputs completely. Sometimes it works fine, but on other occasions you just end up with a grey screen in the VNC client and no output on the HDMI.

I think this is an effect of the recent changes - the older versions before the code was refactored didn't do it, and rolling back to one of them seems to fix the problem. I'll try and work out which change is causing the problem.

**Comment by sshock** [comment]
I ran into this issue with a new raspberry pi 5, because `LABWC_FALLBACK_OUTPUT`, although set in `/etc/xdg/labwc/environment`, was not being set in `/etc/xdg/labwc-greeter/environment`.  After adding it to the latter and rebooting, problem is solved!

**Comment by spl237** [comment]
> I ran into this issue with a new raspberry pi 5, because `LABWC_FALLBACK_OUTPUT`, although set in `/etc/xdg/labwc/environment`, was not being set in `/etc/xdg/labwc-greeter/environment`. After adding it to the latter and rebooting, problem is solved!

Well-spotted - I'd forgotten to do that. I'll add it to the standard configuration.

**Comment by sshock** [comment]
Ok, so yeah, I just checked and there is an `environment.dpkg-dist` in the greeter, and in that one the LABWC_FALLBACK_OUTPUT is present.

In comparing that to mine, another difference I see is my XKB_DEFAULT_LAYOUT is "us", whereas the distro one is "gb".

So here's what I'm guessing happened:
* The image that came on the SD card (that I got from CanaKit) probably already included the greeter (`raspberrypi-ui-mods`), but probably an older version without this fix.
* Very early on during first boot-up it allowed me to select language and keyboard.
* I probably chose "us" and that probably modified the environment config file.
* Then it went through and did  updates.

I'm used to doing updates with Debian, and for any modified config files under /etc it gives you the option to use your version, distro package version, or inspect side by side and fix manually.  But as raspberry pi is trying to make things user-friendly, they probably passed in an option to do updates non-interactively, and default to keeping any local changes.
